scapy/contrib/gxrp.py

```diff
@@ -0,0 +1,228 @@
+        ),
+    ]
+
+    def do_dissect(self, s):
+        s = super(GARP_ATTRIBUTE, self).do_dissect(s)
+        if self.len is not None and self.event == 0 and self.len > 2:
+            warning("Non-empty payload at LeaveAll event")
+        return s
+
+    def extract_padding(self, s):
+        boundary = self.len - 2
+        return s[:boundary], s[boundary:]
+
+    def guess_payload_class(self, payload):
+        try:
+            garp_message = self.parent
+            garp = garp_message.parent
+            llc = garp.underlayer
+            dot3 = llc.underlayer
+            if (
+                dot3.dst == "01:80:c2:00:00:21"
+            ):  # IEEE802.1D-2004 12.4 Table 12-1
+                return GVRP
+            elif (
+                dot3.dst == "01:80:c2:00:00:20"
+            ):  # IEEE802.1D-2004 12.4 Table 12-1
+                if garp_message.type == 1:  # IEEE802.1D-2004 10.3.1.3
+                    return GMRP_GROUP
+                elif garp_message.type == 2:  # IEEE802.1D-2004 10.3.1.3
+                    return GMRP_SERVICE
+        except AttributeError:
+            pass
+        Packet.guess_payload_class(self, payload)
+
+
+def parse_next_attr(pkt, lst, cur, remain):
+    if len(remain) == 0 or remain[0:1] == b"\x00":  # IEEE802.1D-2004 12.10.2.7
+        return None
+    elif remain[0] >= 2:  # minimal attribute size
+        return GARP_ATTRIBUTE
+    else:
+        return None
+
+
+class GARP_MESSAGE(Packet):
+    """
+    GARP message container
+    """
+
+    name = "GARP Message"
+    fields_desc = [
+        ByteField("type", 0x01),
+        PacketListField("attrs", [], next_cls_cb=parse_next_attr),
+        ByteField("end_mark", 0x0),
+    ]
+
+    def extract_padding(self, s):
+        return b"", s
+
+
+def parse_next_msg(pkt, lst, cur, remain):
+    if len(remain) == 0 or remain[0:1] == b"\x00":  # IEEE802.1D-2004 12.10.2.7
+        return None
+    else:
+        return GARP_MESSAGE
+
+
+class GARP(Packet):
+    """
+    GARP packet
+    """
+
+    name = "GARP"
+
+    fields_desc = [
+        ShortField("proto_id", 0x0001),  # IEEE802.1D-2004 12.10.2.1
+        PacketListField("msgs", [], next_cls_cb=parse_next_msg),
+        ByteField("end_mark", 0x0),
+    ]  # IEEE802.1D-2004 12.10.2.7
+
+
+class LLC_GARP(LLC):
+    """
+    Dummy class for layer binding
+    """
+
+    payload_guess = []
+
+
+split_layers(Dot3, LLC)
+[
+    bind_layers(Dot3, LLC_GARP, dst=mac)
+    for mac in [
+        "01:80:c2:00:00:20",
+        "01:80:c2:00:00:21",
+        "01:80:c2:00:00:22",
+        "01:80:c2:00:00:23",
+        "01:80:c2:00:00:24",
+        "01:80:c2:00:00:25",
+        "01:80:c2:00:00:26",
+        "01:80:c2:00:00:27",
+        "01:80:c2:00:00:28",
+        "01:80:c2:00:00:29",
+        "01:80:c2:00:00:2a",
+        "01:80:c2:00:00:2b",
+        "01:80:c2:00:00:2c",
+        "01:80:c2:00:00:2d",
+        "01:80:c2:00:00:2e",
+        "01:80:c2:00:00:2f",
+    ]
+]  # IEEE802.1D-2004 12.4 Table 12-1
```

Comment:
This looks a bit funky :sweat_smile:
```
for mac in [...]:
    bind_layers(...)
```
sounds better?

Reply:
done